--- /etc/inc/pkg-utils.orig.inc 2023-06-20 01:48:04.989690000 +0000
+++ /etc/inc/pkg-utils.inc      2023-06-20 01:53:07.137712000 +0000
@@ -387,7 +387,7 @@
 	if ($base_packages) {
 		$repo_param = "";
 	} else {
-		$repo_param = "-r {$g['product_name']}";
+		$repo_param = "";
 	}

 	/*
@@ -486,7 +486,7 @@
 			    $err);
 			if (!$base_packages &&
 			    rtrim($out) != g_get('product_name')) {
-				continue;
+				//continue with unofficial packages;
 			}

 			$pkg_info['installed'] = true;
