--- /etc/inc/pkg-utils.inc      2021-02-18 15:26:41.486078000 -0300
+++ /root/pkg-utils.inc 2021-02-18 11:18:07.356449000 -0300
@@ -402,7 +402,7 @@
        if ($base_packages) {
                $repo_param = "";
        } else {
-               $repo_param = "";
+               $repo_param = "-r {$g['product_name']}";
        }

        /*
@@ -499,7 +499,7 @@
                            $err);
                        if (!$base_packages &&
                            rtrim($out) != $g['product_name']) {
-                               //continue;
+                               continue;
                        }

                        $pkg_info['installed'] = true;
