--- /usr/local/www/pkg.orig.php	2017-04-05 17:18:25.349676000 -0300
+++ /usr/local/www/pkg.php	2017-04-05 17:20:49.204578000 -0300
@@ -324,6 +324,7 @@
 
 <form action="pkg.php" name="pkgform" method="get">
 	<input type='hidden' name='xml' value='<?=$_REQUEST['xml']?>' />
+		<div id="submenu"> </div>
 		<div id="mainarea" class="panel panel-default">
 			<table id="mainarea" class="table table-striped table-hover table-condensed">
 				<thead>
